libsignal-protocol/src/crypto/mod.rs

```diff
@@ -0,0 +1,381 @@
+#[cfg(feature = "crypto-native")]
+mod native;
+#[cfg(feature = "crypto-native")]
+pub use self::native::DefaultCrypto;
+
+#[cfg(feature = "crypto-openssl")]
+mod openssl;
+#[cfg(feature = "crypto-openssl")]
+pub use self::openssl::OpenSSLCrypto;
+
+use std::{
+    cell::RefCell,
+    convert::TryFrom,
+    os::raw::{c_int, c_void},
+    pin::Pin,
+    slice,
+};
+use sys::{signal_buffer, signal_crypto_provider};
+
+use crate::{
+    buffer::Buffer,
+    errors::{InternalError, IntoInternalErrorCode},
+};
+
+#[derive(Debug, Clone)]
+pub struct SignalCipherTypeError(i32);
+
+#[derive(Copy, Clone)]
+pub enum CipherMode {
+    Encrypt,
+    Decrypt,
+}
+pub enum SignalCipherType {
+    AesCtrNoPadding,
+    AesCbcPkcs5,
+}
+
+impl TryFrom<i32> for SignalCipherType {
+    type Error = SignalCipherTypeError;
+
+    #[inline]
+    fn try_from(v: i32) -> Result<Self, Self::Error> {
+        match v as u32 {
+            sys::SG_CIPHER_AES_CTR_NOPADDING => {
+                Ok(SignalCipherType::AesCtrNoPadding)
+            },
+            sys::SG_CIPHER_AES_CBC_PKCS5 => Ok(SignalCipherType::AesCbcPkcs5),
+            _ => Err(SignalCipherTypeError(v)),
+        }
+    }
+}
+
+/// Something which can calculate a SHA-256 HMAC.
+pub trait Sha256Hmac {
+    fn update(&mut self, data: &[u8]) -> Result<(), InternalError>;
+    fn finalize(&mut self) -> Result<Vec<u8>, InternalError>;
+}
+
+/// Something which can generate a SHA-512 hash.
+pub trait Sha512Digest {
+    fn update(&mut self, data: &[u8]) -> Result<(), InternalError>;
+    fn finalize(&mut self) -> Result<Vec<u8>, InternalError>;
+}
+
+/// Cryptography routines used in the signal protocol.
+pub trait Crypto {
```

Comment:
also having one method to init and return one structure e.g `Sha256Hmac` or `Sha512Digest` remove the confusion behind all these methods and what order will they be executed, which is a nice thing IMO.